src/host/readDataCooked.cpp

```diff
@@ -436,633 +201,1024 @@ bool COOKED_READ_DATA::AtEol() const noexcept
+    {
+        _popupDone();
+    }
+
+    const auto& textBuffer = _screenInfo.GetTextBuffer();
+    const auto eraseDistance = (_coordEnd.y - _coordBeg.y) * textBuffer.GetSize().Width() + _coordEnd.x - _coordBeg.x;
+
+    std::ignore = _screenInfo.SetCursorPosition(_coordBeg, true);
+    _erase(eraseDistance);
+
+    std::ignore = _screenInfo.SetCursorPosition(_coordBeg, true);
+    _coordEnd = _coordBeg;
+    _coordCursor = _coordBeg;
 }
 
-// Routine Description:
-// - This method handles the various actions that occur on the edit line like pressing keys left/right/up/down, paging, and
-//   the final ENTER key press that will end the wait and finally return the data.
-// Arguments:
-// - pCookedReadData - Pointer to cooked read data information (edit line, client buffer, etc.)
-// - wch - The most recently pressed/retrieved character from the input buffer (keystroke)
-// - keyState - Modifier keys/state information with the pressed key/character
-// - status - The return code to pass to the client
-// Return Value:
-// - true if read is completed. false if we need to keep waiting and be called again with the user's next keystroke.
-bool COOKED_READ_DATA::ProcessInput(const wchar_t wchOrig,
-                                    const DWORD keyState,
-                                    NTSTATUS& status)
+// The counter-part to EraseBeforeResize().
+void COOKED_READ_DATA::RedrawAfterResize()
 {
-    const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-    size_t NumSpaces = 0;
-    til::CoordType ScrollY = 0;
-    size_t NumToWrite;
-    auto wch = wchOrig;
-    bool fStartFromDelim;
-
-    status = STATUS_SUCCESS;
-    if (_bytesRead >= (_bufferSize - (2 * sizeof(WCHAR))) && wch != UNICODE_CARRIAGERETURN && wch != UNICODE_BACKSPACE)
-    {
-        return false;
-    }
+    _coordBeg = _screenInfo.GetTextBuffer().GetCursor().GetPosition();
+    _coordEnd = _coordBeg;
+    _coordCursor = _coordBeg;
 
-    if (_ctrlWakeupMask != 0 && wch < L' ' && (_ctrlWakeupMask & (1 << wch)))
+    _markAsDirty();
+    _flushBuffer();
+}
+
+void COOKED_READ_DATA::SetInsertMode(bool insertMode) noexcept
+{
+    _insertMode = insertMode;
+}
+
+bool COOKED_READ_DATA::IsEmpty() const noexcept
+{
+    return _buffer.empty() && _popups.empty();
+}
+
+bool COOKED_READ_DATA::PresentingPopup() const noexcept
+{
+    return !_popups.empty();
+}
+
+til::point_span COOKED_READ_DATA::GetBoundaries() const noexcept
+{
+    return { _coordBeg, _coordEnd };
+}
+
+// _wordPrev and _wordNext implement the classic Windows word-wise cursor movement algorithm, as traditionally used by
+// conhost, notepad, Visual Studio and other "old" applications. If you look closely you can see how they're the exact
+// same "skip 1 char, skip x, skip not-x", but since the "x" between them is different (non-words for _wordPrev and
+// words for _wordNext) it results in the inconsistent feeling that these have compared to more modern algorithms.

```

Comment:
Presumably this is the thing you were talking about in #15787? We could leave a `// TODO: GH#15787` here if you want